Building an object detection application with tensorflow.js and react.js on Ubuntu 20.04 (#2135)
section-engineering-education/engineering-education: articles/building-an-object-detection-application-with-tensorfow.js-and-react.js/index.md

We will install TensorFlow and the pre-built model into our application by issuing the following commands:

```bash
    import * as tf from "@tensorflow/tfjs";
    import * as cocossd from "@tensorflow-models/coco-ssd";
```

The first command will first import the TensorFlow module into our application. The second command then imports the COCO SSD model from the TensorFlow model.

### Building the React.js app

We first import all the dependencies on the 'App.js' folder, including the React.js library, TensorFlow.js, and the COCO SSD pre-trained model. It imports the required dependencies into our application. 

```bash
import React, { useRef, useState, useEffect } from "react";
import * as tf from "@tensorflow/tfjs";
import * as cocossd from "@tensorflow-models/coco-ssd";
import Webcam from "react-webcam";
import "./App.css";
import { drawRect } from "./utilities";
```

The next thing that we are going to do is to load our network into our model. This is achieved by issuing the following command:

```bash
    const net = await cocossd.load();
```

> We first create a new variable called 'net'. Because our function is asynchronous, we wait for the 'cocossd' model to load.

Reviewer: What would happen if our function would be of a synchronous paradigm? What is the significance of the asynchronous function in this context?

Finally, once the objects have been detected in the webcam video, we can draw bounding boxes to show where these objects are being detected and assign class labels. This is accomplished by using the function below in the 'utilities.js' folder.

```bash
    export const drawRect = (detections, ctx) =>{
    // Loop through each prediction
    detections.forEach(prediction => {
  
      // Extract boxes and classes
      const [x, y, width, height] = prediction['bbox']; 
      const text = prediction['class']; 
  
      // Set styling
      const color = Math.floor(Math.random()*16777215).toString(16);
      ctx.strokeStyle = '#' + color
      ctx.font = '18px Arial';
  
      // Draw rectangles and text
      ctx.beginPath();   
      ctx.fillStyle = '#' + color
      ctx.fillText(text, x, y);
      ctx.rect(x, y, width, height); 
      ctx.stroke();
    });
  }
```
This is then imported into the main 'App.js' file using the command:

```bash
    import { drawRect } from "./utilities";
```

It is then executed with the command:

```bash
    drawRect(obj, ctx);
```
This enables bounding boxes to be drawn on the video.

> To access the complete code, try and test the code out for yourself, you can find the complete code [here](https://replit.com/@wmkinyan/RealTimeObjectDetectionTFJSReact#README.md).

Reviewer: ```suggestion
To access and tinker with the complete block of code, you can find it [here](https://replit.com/@wmkinyan/RealTimeObjectDetectionTFJSReact#README.md).
```

### Conclusion

> That wraps it all! 
> 
> In this tutorial, we have learned how to build an object detection application using the TensorFlow library in the React app.

Reviewer: ```suggestion
That wraps it up! In this tutorial, we have learned how to build an object detection application using the TensorFlow library in the React app.
```
